### DISCUSSION TITLE
Best approach to profiling queries that are outliers?

### DISCUSSION THREAD

**Discussion by trevhunter** [ravendb/ravendb#12925] [Q&A] [answered]
Overall, we're very happy with the performance of RavenDB for our use cases, but we're always interested in learning more ways to optimize. One area we'd like to understand more about is how to best profile requests that are *outliers in performance* .

## Scenario
* Around 100 Requests per second hitting one node.
* Simple query against an auto index (`from x where x.someproperty == $p0 order by someotherproperty`)
* Most results are empty (nothing matched the criteria)
* End to End SLO p95 of 150ms

Taking a brief window of traffic watch from the RavenDB server gives us a histogram like this:
![image](https://user-images.githubusercontent.com/3181772/138952191-16955e3a-539f-466f-8b07-a4286bee7f78.png)

As you can see, we're meeting our SLO (at least from the RavenDB response times, we still need to add on our app processing times to this). 476 out of 552 requests are under 50ms (in fact, 338 requests are under 10ms). Those are cool, but not interesting at this point. I'm more interested in understanding the cause of the outliers > 150ms. Diving into these, you can see that some of them go up into the 600ms range. 

![image](https://user-images.githubusercontent.com/3181772/138952725-c6f3c694-86dc-44b0-b6a2-788fa488e0c3.png)

Looking at these outlier requests, there's nothing special about them in the traffic watch - no big result set, no change in query criteria. We realize that there's a tonne of ways that cause these - from cache misses to GC pauses, to maybe other processes on the machine to even client delays reading the results etc. 

So the question is - how do we best profile these outliers in order to make informed decisions about further finetuning to minimize them? Traffic watch is a great way to identify them, but it doesn't help inform us if we should be adding more memory to the node, or if there are other aspects we could take a look at. Note: Re-running the query later doesn't usually show the same performance profile for those same parameter values.

**Comment by ayende** [comment, marked as answer]
Easiest place to start would be to register a `OnBeforeQuery` and add `include timings()`, then you have the server details on what was costly.
I also added: https://issues.hibernatingrhinos.com/issue/RavenDB-17401

If you repeat the slow queries, you're likely to see that they are fast now, which leads to the curse of trying to optimize the P99. 
The problem is that a query may need to hit the disk to read a document, which incur an I/O cost, the second time, you won't see that.

Typically for such scenarios we are actually watching the server side metrics as well, trying to see IO wait, CPU spikes, GC events, etc. 
*really* hard to correlate specifically to a single query, unless it is doing something expensive directly.

**Reply by trevhunter** [reply]
Thanks @ayende - the timings() aspect is a good place to start as at least there we should see a hint if its storage/cache related or not. The enhancement to traffic watch would also be most welcome to do it all within the studio instead of client code. 

Yes, it's very hard to correlate these outliers with other things going on in the server / environment, and in general it's a lower priority as long as we're inside our SLO, but the more ways we can trace/instrument the execution of a single request, the more informed we may get with some experimentation (e.g. adding more memory etc.)
